libs/wire-api/src/Wire/API/Team/Feature.hs

```diff
@@ -602,40 +536,24 @@ instance ToSchema LockStatusResponse where
       LockStatusResponse
         <$> _unlockStatus .= field "lockStatus" schema
 
-newtype ImplicitLockStatus (cfg :: Type) = ImplicitLockStatus {_unImplicitLockStatus :: WithStatus cfg}
+newtype ImplicitLockStatus (cfg :: Type) = ImplicitLockStatus {_unImplicitLockStatus :: LockableFeature cfg}
   deriving newtype (Eq, Show, Arbitrary)
 
 instance (IsFeatureConfig a, ToSchema a) => ToJSON (ImplicitLockStatus a) where
   toJSON (ImplicitLockStatus a) = A.toJSON $ forgetLock a
 
 instance (IsFeatureConfig a, ToSchema a) => FromJSON (ImplicitLockStatus a) where
-  parseJSON v = ImplicitLockStatus . withLockStatus (wsLockStatus $ defFeatureStatus @a) <$> A.parseJSON v
+  parseJSON v = ImplicitLockStatus . withLockStatus ((def @(LockableFeature a)).lockStatus) <$> A.parseJSON v
 
 -- | Convert a feature coming from the database to its public form. This can be
 -- overridden on a feature basis by implementing the `computeFeature` method of
 -- the `GetFeatureConfig` class.
-genericComputeFeature ::
-  WithStatus cfg ->
-  Maybe LockStatus ->
-  DbFeature cfg ->
-  WithStatus cfg
-genericComputeFeature defFeature lockStatus dbFeature =
-  case fromMaybe (wsLockStatus defFeature) lockStatus of
-    LockStatusLocked -> setLockStatus LockStatusLocked defFeature
-    LockStatusUnlocked -> withUnlocked $ unDbFeature dbFeature (forgetLock defFeature)
-
--- | This contains the pure business logic for users from teams
-computeFeatureConfigForTeamUser :: Maybe (WithStatusNoLock cfg) -> Maybe LockStatus -> WithStatus cfg -> WithStatus cfg
-computeFeatureConfigForTeamUser mStatusDb mLockStatusDb defStatus =
-  case lockStatus of
-    LockStatusLocked ->
-      withLocked (forgetLock defStatus)
-    LockStatusUnlocked ->
-      withUnlocked $ case mStatusDb of
-        Nothing -> forgetLock defStatus
-        Just fs -> fs
-  where
-    lockStatus = fromMaybe (wsLockStatus defStatus) mLockStatusDb
+genericComputeFeature :: forall cfg. LockableFeature cfg -> DbFeature cfg -> LockableFeature cfg
+genericComputeFeature defFeature dbFeature =
+  let feat = applyDbFeature dbFeature defFeature
+   in case feat.lockStatus of
+        LockStatusLocked -> defFeature {lockStatus = LockStatusLocked}
+        LockStatusUnlocked -> feat
```

Comment:
Maybe just a heads up to other reviewers looking at this (besides me), this logic needs to be equivalent to the old one. I think it makes sense like @pcapriotti explained, but yeah, more eyes can't hurt.